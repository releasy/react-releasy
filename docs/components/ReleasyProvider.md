# ReleasyProvider

The root component.

## Usage

First create an instance of `Config`:

```javascript
import { Config, Link } from 'react-releasy';

const link = new Link({
  url: 'https://yourserveraddress.com/graphql',
});

const config = new Config({
  link,
});
```

Then wrap the application:

```javascript
import { ReleasyProvider } from 'react-releasy';

ReactDOM.render(
  <ReleasyProvider config={config}>
    <MyApplication />
  </ReleasyProvider>,
  document.getElementById('root'),
);
```

After this we can inject the `Relay` `environment` into our components using [withReleasy](../hocs/withReleasy.md), or even getting it directly:

```javascript
import { getEnvironment } from 'react-releasy';


const environment = getEnvironment();
```

## Props

- **config**: see [Config](../classes/Config.md)

----

[Go back](../)
