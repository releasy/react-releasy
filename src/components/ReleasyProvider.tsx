import * as React from 'react';
import invariant from 'invariant';

import GlobalContext, {
  GlobalContextValueType,
  setConfig,
  setEnvironment,
} from '../contexts/GlobalContext';
import Config from '../core/Config';
import createEnvironment from '../relay/createEnvironment';

type ReleasyProviderPropsType = {
  config: Config,
  children: React.ReactNode,
};

type ReleasyProviderStateType = {
  globalContextValue: GlobalContextValueType,
};

class ReleasyProvider extends React.PureComponent<
  ReleasyProviderPropsType,
  ReleasyProviderStateType
> {
  constructor(props: ReleasyProviderPropsType) {
    super(props);

    invariant(props.config, 'ReleasyProvider: missing "config"');

    const { config } = props;
    const environment = createEnvironment(config);

    setConfig(config);
    setEnvironment(environment);

    this.state = {
      globalContextValue: {
        config,
        environment,
      },
    };
  }

  render() {
    const { children } = this.props;
    const { globalContextValue } = this.state;

    return (
      <GlobalContext.Provider value={globalContextValue}>
        {children}
      </GlobalContext.Provider>
    );
  }
}

export default ReleasyProvider;
