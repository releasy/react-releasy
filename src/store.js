// @flow
import { Environment } from 'relay-runtime';

const ENVIRONMENT_KEY = 'environment';

const store: Map<string, Environment> = new Map();

export const getEnvironment = (): Environment => store.get(ENVIRONMENT_KEY);

export const setEnvironment = (environment: Environment): void => {
  store.set(ENVIRONMENT_KEY, environment);
};

export default store;
