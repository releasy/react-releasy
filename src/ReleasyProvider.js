// @flow
import * as React from 'react';
import * as PropTypes from 'prop-types';
import { Environment } from 'relay-runtime';
import invariant from 'invariant';

import { getEnvironment } from './store';

type Props = {
  children: React.Node,
};

class ReleasyProvider extends React.PureComponent<Props> {
  static propTypes = {
    children: PropTypes.element.isRequired,
  };

  static childContextTypes = {
    environment: PropTypes.instanceOf(Environment).isRequired,
  };

  constructor(props: Props) {
    invariant(props.children, 'ReleasyProvider: missing "children"');

    super(props);
  }

  getChildContext() {
    return {
      environment: getEnvironment(),
    };
  }

  render() {
    return this.props.children;
  }
}

export default ReleasyProvider;
