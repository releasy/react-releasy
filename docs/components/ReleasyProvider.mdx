---
name: ReleasyProvider
menu: Components
route: /components/releasy-provider
---

import { PropsTable } from 'docz'
import { ReleasyProvider } from '../../src'

# ReleasyProvider

## Properties

<PropsTable of={ReleasyProvider} />