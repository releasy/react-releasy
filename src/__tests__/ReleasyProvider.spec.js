import React from 'react';
import { shallow } from 'enzyme';

import DumbComponent from '../../test/components/DumbComponent';
import createConfig from '../../test/fixtures/createConfig';

import ReleasyProvider from '../ReleasyProvider';
import { getEnvironment } from '../store';

it('should render a valid ReleasyProvider', () => {
  const config = createConfig();

  const component = shallow(
    <ReleasyProvider config={config}>
      <DumbComponent />
    </ReleasyProvider>,
  );

  expect(component.instance()).toMatchSnapshot();
});

it('ReleasyProvider should have a valid getChildContext()', () => {
  const config = createConfig();

  const component = shallow(
    <ReleasyProvider config={config}>
      <DumbComponent />
    </ReleasyProvider>,
  );

  expect(component.instance().getChildContext()).toMatchSnapshot();
});

it('should thrown an error without "config"', () => {
  const renderProvider = () => {
    shallow(
      <ReleasyProvider>
        <DumbComponent />
      </ReleasyProvider>,
    );
  };

  expect(renderProvider).toThrowErrorMatchingSnapshot();
});

it('should thrown an error without "children"', () => {
  const renderProvider = () => {
    const config = createConfig();

    shallow(<ReleasyProvider config={config} />);
  };

  expect(renderProvider).toThrowErrorMatchingSnapshot();
});


it('should be capable to get environment statically', () => {
  const config = createConfig();

  const component = shallow(
    <ReleasyProvider config={config}>
      <DumbComponent />
    </ReleasyProvider>,
  );

  expect(getEnvironment()).toMatchSnapshot();
});
