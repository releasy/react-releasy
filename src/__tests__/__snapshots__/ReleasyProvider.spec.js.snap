// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReleasyProvider should have a valid getChildContext() 1`] = `
Object {
  "environment": RelayModernEnvironment {
    "__setNet": [Function],
    "_deferrableSelections": Set {},
    "_network": Object {
      "execute": [Function],
    },
    "_publishQueue": RelayPublishQueue {
      "_appliedOptimisticUpdates": Set {},
      "_backup": Object {},
      "_handlerProvider": [Function],
      "_pendingBackupRebase": false,
      "_pendingData": Set {},
      "_pendingOptimisticUpdates": Set {},
      "_pendingUpdaters": Set {},
      "_store": RelayMarkSweepStore {
        "_gcEnabled": true,
        "_hasScheduledGC": false,
        "_index": 0,
        "_recordSource": Object {},
        "_roots": Map {},
        "_subscriptions": Set {},
        "_updatedRecordIDs": Object {},
      },
    },
    "_store": RelayMarkSweepStore {
      "_gcEnabled": true,
      "_hasScheduledGC": false,
      "_index": 0,
      "_recordSource": Object {},
      "_roots": Map {},
      "_subscriptions": Set {},
      "_updatedRecordIDs": Object {},
    },
    "configName": undefined,
    "unstable_internal": Object {
      "areEqualSelectors": [Function],
      "createFragmentSpecResolver": [Function],
      "createOperationSelector": [Function],
      "getDataIDsFromObject": [Function],
      "getFragment": [Function],
      "getRequest": [Function],
      "getSelector": [Function],
      "getSelectorList": [Function],
      "getSelectorsFromObject": [Function],
      "getVariablesFromObject": [Function],
    },
  },
}
`;

exports[`should be capable to get environment statically 1`] = `
RelayModernEnvironment {
  "__setNet": [Function],
  "_deferrableSelections": Set {},
  "_network": Object {
    "execute": [Function],
  },
  "_publishQueue": RelayPublishQueue {
    "_appliedOptimisticUpdates": Set {},
    "_backup": Object {},
    "_handlerProvider": [Function],
    "_pendingBackupRebase": false,
    "_pendingData": Set {},
    "_pendingOptimisticUpdates": Set {},
    "_pendingUpdaters": Set {},
    "_store": RelayMarkSweepStore {
      "_gcEnabled": true,
      "_hasScheduledGC": false,
      "_index": 0,
      "_recordSource": Object {},
      "_roots": Map {},
      "_subscriptions": Set {},
      "_updatedRecordIDs": Object {},
    },
  },
  "_store": RelayMarkSweepStore {
    "_gcEnabled": true,
    "_hasScheduledGC": false,
    "_index": 0,
    "_recordSource": Object {},
    "_roots": Map {},
    "_subscriptions": Set {},
    "_updatedRecordIDs": Object {},
  },
  "configName": undefined,
  "unstable_internal": Object {
    "areEqualSelectors": [Function],
    "createFragmentSpecResolver": [Function],
    "createOperationSelector": [Function],
    "getDataIDsFromObject": [Function],
    "getFragment": [Function],
    "getRequest": [Function],
    "getSelector": [Function],
    "getSelectorList": [Function],
    "getSelectorsFromObject": [Function],
    "getVariablesFromObject": [Function],
  },
}
`;

exports[`should render a valid ReleasyProvider 1`] = `
ReleasyProvider {
  "context": Object {},
  "props": Object {
    "children": <DumbComponent />,
    "config": Config {
      "cache": null,
      "devTools": false,
      "link": Link {
        "fetch": [Function],
        "fetchTimeout": 30000,
        "headers": Object {},
        "retryDelays": Array [
          1000,
          3000,
          5000,
          10000,
        ],
        "url": "/graphql",
      },
      "networkLogger": false,
    },
  },
  "refs": Object {},
  "state": Object {},
  "updater": Updater {
    "_callbacks": Array [],
    "_renderer": ReactShallowRenderer {
      "_context": Object {},
      "_element": <ReleasyProvider
        config={
          Config {
            "cache": null,
            "devTools": false,
            "link": Link {
              "fetch": [Function],
              "fetchTimeout": 30000,
              "headers": Object {},
              "retryDelays": Array [
                1000,
                3000,
                5000,
                10000,
              ],
              "url": "/graphql",
            },
            "networkLogger": false,
          }
        }
      >
        <DumbComponent />
      </ReleasyProvider>,
      "_forcedUpdate": false,
      "_instance": [Circular],
      "_newState": null,
      "_rendered": <DumbComponent />,
      "_rendering": false,
      "_updater": [Circular],
    },
  },
}
`;

exports[`should thrown an error without "children" 1`] = `"ReleasyProvider: missing \\"children\\""`;

exports[`should thrown an error without "config" 1`] = `"ReleasyProvider: missing \\"config\\""`;
